keps/sig-auth/2718-20210511-client-exec-proxy/README.md

```diff
@@ -0,0 +1,811 @@
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+This proposal is to implement an extension in Kubernetes client libraries to
+forward requests to a proxy process for request signing.  The proxy will
+receive an HTTP request and will have the opportunity to sign the request,
+attach headers, use hardware protected keys for mTLS, and/or modify the
+request.  The final request will be constructed and forwarded to the apiserver.
+
+It is already possible, by setting the appropriate HTTP_PROXY, HTTPS_PROXY
+environment variables, to configure a Kubernetes client to send requests to a
+proxy.  While this may work in some situations, it is not ideal for a generic
+request signing solution across all client configurations.  Firstly, http
+proxies are already commonly integrated in client networking configurations, so
+for these clients setting up an additional proxy configuration might not be
+possible, or at least not straightforward.  Secondly, this approach requires a
+long running external proxy process which introduces operational complexity.
+Thirdly, communicating with a proxy for request signing requires a secure
+communication channel between the client and the proxy, which means in a case
+where multiple users share a client computer (for example, a shared bastion),
+configuring and securing a local proxy to prevent signing requests as a
+different user becomes more complicated.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Proxy Proposal
+
+The proxy exec plugin is a long-running proxy process which is executed by the
+client on first use.  The proxy binary location is defined in the `ExecConfig`
```

Comment:
Do we have one proxy per process, one per client, one per config, etc...? How does the caching work?

Reply:
Currently thinking one proxy per process, where the connection to the proxy would be cached based on the input parameters passed over stdin (which includes the cluster information). Added a note to that effect.